pyccl/halos/halo_model.py

```diff
@@ -1267,138 +1291,127 @@ def halomod_Tk3D_SSC(cosmo, hmc,
-    elif (p_of_k_a is None) or (str(p_of_k_a) == 'linear'):
+    elif (p_of_k_a is None) or str(p_of_k_a) == 'linear':
         pk2d = cosmo.get_linear_power('delta_matter:delta_matter')
     elif str(p_of_k_a) == 'nonlinear':
         pk2d = cosmo.get_nonlin_power('delta_matter:delta_matter')
     else:
-        raise TypeError("p_of_k_a must be `None`, \'linear\', "
-                        "\'nonlinear\' or a `Pk2D` object")
+        raise ValueError("p_of_k_a must be `None`, 'linear', "
+                         "'nonlinear' or a `Pk2D` object")
 
     def get_norm(normprof, prof, sf):
-        if normprof:
-            return hmc.profile_norm(cosmo, sf, prof)
-        else:
-            return 1
+        return hmc.profile_norm(cosmo, sf, prof) if normprof else 1
 
-    na = len(a_arr)
-    nk = len(k_use)
-    dpk12 = np.zeros([na, nk])
-    dpk34 = np.zeros([na, nk])
+    dpk12, dpk34 = [np.zeros((len(a_arr), len(k_use))) for _ in range(2)]
     for ia, aa in enumerate(a_arr):
         # Compute profile normalizations
         norm1 = get_norm(normprof1, prof1, aa)
         i11_1 = hmc.I_1_1(cosmo, k_use, aa, prof1)
         # Compute second profile normalization
-        if prof2 is None:
+        if prof2 == prof1:
             norm2 = norm1
             i11_2 = i11_1
         else:
             norm2 = get_norm(normprof2, prof2, aa)
             i11_2 = hmc.I_1_1(cosmo, k_use, aa, prof2)
-        if prof3 is None:
+
+        if prof3 == prof1:
             norm3 = norm1
             i11_3 = i11_1
         else:
             norm3 = get_norm(normprof3, prof3, aa)
             i11_3 = hmc.I_1_1(cosmo, k_use, aa, prof3)
-        if prof4 is None:
-            norm4 = norm3
-            i11_4 = i11_3
+
+        if prof4 == prof2:
+            norm4 = norm2
+            i11_4 = i11_2
         else:
             norm4 = get_norm(normprof4, prof4, aa)
             i11_4 = hmc.I_1_1(cosmo, k_use, aa, prof4)
 
-        i12_12 = hmc.I_1_2(cosmo, k_use, aa, prof1,
-                           prof12_2pt, prof2)
-        if (prof3 is None) and (prof4 is None) and (prof34_2pt is None):
+        i12_12 = hmc.I_1_2(cosmo, k_use, aa, prof1, prof12_2pt, prof2)
+
+        if (prof3, prof4) == (prof1, prof2):
             i12_34 = i12_12
         else:
-            i12_34 = hmc.I_1_2(cosmo, k_use, aa, prof3_bak,
-                               prof34_2pt_bak, prof4)
+            i12_34 = hmc.I_1_2(cosmo, k_use, aa, prof3, prof34_2pt, prof4)
+
         norm12 = norm1 * norm2
         norm34 = norm3 * norm4
 
         pk = pk2d.eval(k_use, aa, cosmo)
         dpk = pk2d.eval_dlogpk_dlogk(k_use, aa, cosmo)
         # (47/21 - 1/3 dlogPk/dlogk) * I11 * I11 * Pk+I12
-        dpk12[ia, :] = norm12*((2.2380952381-dpk/3)*i11_1*i11_2*pk+i12_12)
-        dpk34[ia, :] = norm34*((2.2380952381-dpk/3)*i11_3*i11_4*pk+i12_34)
+        dpk12[ia, :] = norm12*((47/21 - dpk/3)*i11_1*i11_2*pk + i12_12)
+        dpk34[ia, :] = norm34*((47/21 - dpk/3)*i11_3*i11_4*pk + i12_34)
```

Comment:
while 47/21 is clearer, it's also slower

Reply:
This is a common misconception among Python users who come from different languages. When the interpreter first runs, it sees all that and calculates whatever can be calculated when creating the `.pyc` files in `__pycache__`. So it isn't slower.